logictest: race when updating DistSQL span resolver

pkg/sql/logictest/tests/fakedist-vec-off/fakedist-vec-off_test.TestLogic_rename_database [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9169654?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9169654?buildTab=artifacts#/) on master @ [87e227e4e952fb788099379cf2881bcb814776fd](https://github.com/cockroachdb/cockroach/commits/87e227e4e952fb788099379cf2881bcb814776fd):


```
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).partitionSpansSystem(0x4000b0af00, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x4005833320, 0x6, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1280 +0x180
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PartitionSpans(0x0?, {0x5d08388?, 0x400836dc20?}, 0x0?, {0x4005833320?, 0x0?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1158 +0x44
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planTableReaders(0x4000b0af00, {0x5d08388, 0x400836dc20}, 0x0?, 0x400798b180, 0x400899a3f8)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1768 +0x68
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createTableReaders(0x522782de79163209?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, 0x400673f500)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1603 +0x1e4
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x7e0a8452681d0785?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a880?, 0x400673f500?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3422 +0x1cc
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForIndexJoin(0x520ffb1a0ffbee97?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, 0x400798af80)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:2618 +0x40
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x61b43a608e015cd8?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a1f0?, 0x400798af80?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3349 +0x138
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x40048d29b8?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a6f8?, 0x4005a192c0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3390 +0x200
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x5da5020?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0ac00?, 0x400836dbf0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3425 +0x910
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x4006bfbff8?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a6f8?, 0x4005a19400?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3390 +0x200
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlan(0x5d08388?, {0x5d08388?, 0x400836dc20?}, 0x1?, {{0x5d0a6f8, 0x4005a19400}, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3297 +0x60
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x4001e7a5e8?, {0x5d08388, 0x400836dc20}, 0xf0?, 0x4003d8aab0, 0x40019aa870?, {{0x5d0a6f8?, 0x4005a19400?}, 0x0?}, 0x40092dc380, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1853 +0xd4
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func2(0x4001e7a5e8, 0xffff78389bd0?, {0x5d08388, 0x400836dc20}, 0x100012400?, 0xffff79140798?, 0xffffa61ac1d8?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1597 +0x84
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0x4000b0af00?, {0x5d08388, 0x400836dc20}, 0x4001e7aa08, 0x4001e7a5e8?, 0x4001e7a5e8, 0x40092dc380, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1600 +0x184
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0x4001e7a000, {0x5d08388?, 0x400836dc20}, 0x4001e7a5e8, 0x0?, {0x5d59f80?, 0x400836d800}, 0x4006bfc488?, 0x40043d6c98)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1795 +0x448
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0x4001e7a000, {0x5d082e0, 0x4004cd08c0}, 0x4001e7a5e8, {0x5d59f80, 0x400836d800})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1378 +0x8b8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0x4001e7a000, {0x5d08388?, 0x400836d7a0?}, {{0x5d31520, 0x400a4ead20}, {0x0, 0x0, 0x0}, {0x4b0914c, 0x9f}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:785 +0x1f94
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x5d08388?, 0x400836d7a0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:137 +0x8c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x5d08388?, {0x5d08388?, 0x400836d7a0?}, {0x5d31520?, 0x400a4ead20?}, 0x460d14?, 0x4001caa578?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2683 +0x280
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x4001e7a000, {0x5d08388, 0x400836d7a0}, {{0x5d31520, 0x400a4ead20}, {0x0, 0x0, 0x0}, {0x4b0914c, 0x9f}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:136 +0x36c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x4001caabb8, 0x4001e7a000, 0x4001caa878, 0x4001caa938, 0x4001caa978, 0x400899f958, 0x400899f948, 0x400899f968)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1992 +0x2e0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x4001e7a000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1997 +0xc88
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x4001e7a000, {0x5d08388, 0x400836d380}, 0x400836c930?, 0x10ae420?, 0x4008763000?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1913 +0x18c
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:172 +0x90
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:171 +0x150
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_rename_database.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25706

```
=== RUN   TestLogic_rename_database
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a1247b5611bdd40627ba469c89851fae/logTestLogic_rename_database3246356192
    test_log_scope.go:79: use -show-logs to present logs inline
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x3 pc=0x1bde258]

goroutine 257850 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*RangeIterator).Seek(0x40019aa878, {0x5d08388, 0x400836dc20}, {0x4000bb6450, 0x15, 0x18}, 0x0)
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/range_iter.go:204 +0x2b8
github.com/cockroachdb/cockroach/pkg/sql/physicalplan.(*spanResolverIterator).Seek(0x40019aa870, {0x5d08388, 0x400836dc20}, {{0x4000bb6450, 0x15, 0x18}, {0x4000bb6480, 0x15, 0x15}}, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/physicalplan/span_resolver.go:251 +0x4d0
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).partitionSpan(0x4006bf7e88?, {0x5d08388, 0x400836dc20}, 0x1?, {{0x4000bb6450, 0x15, 0x18}, {0x4000bb6480, 0x15, 0x15}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1201 +0x1cc
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).partitionSpansSystem(0x4000b0af00, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x4005833320, 0x6, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1280 +0x180
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PartitionSpans(0x0?, {0x5d08388?, 0x400836dc20?}, 0x0?, {0x4005833320?, 0x0?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1158 +0x44
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planTableReaders(0x4000b0af00, {0x5d08388, 0x400836dc20}, 0x0?, 0x400798b180, 0x400899a3f8)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1768 +0x68
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createTableReaders(0x522782de79163209?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, 0x400673f500)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1603 +0x1e4
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x7e0a8452681d0785?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a880?, 0x400673f500?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3422 +0x1cc
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForIndexJoin(0x520ffb1a0ffbee97?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, 0x400798af80)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:2618 +0x40
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x61b43a608e015cd8?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a1f0?, 0x400798af80?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3349 +0x138
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x40048d29b8?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a6f8?, 0x4005a192c0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3390 +0x200
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x5da5020?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0ac00?, 0x400836dbf0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3425 +0x910
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlanForPlanNode(0x4006bfbff8?, {0x5d08388, 0x400836dc20}, 0x4003d8aab0, {0x5d0a6f8?, 0x4005a19400?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3390 +0x200
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPhysPlan(0x5d08388?, {0x5d08388?, 0x400836dc20?}, 0x1?, {{0x5d0a6f8, 0x4005a19400}, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:3297 +0x60
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x4001e7a5e8?, {0x5d08388, 0x400836dc20}, 0xf0?, 0x4003d8aab0, 0x40019aa870?, {{0x5d0a6f8?, 0x4005a19400?}, 0x0?}, 0x40092dc380, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1853 +0xd4
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func2(0x4001e7a5e8, 0xffff78389bd0?, {0x5d08388, 0x400836dc20}, 0x100012400?, 0xffff79140798?, 0xffffa61ac1d8?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1597 +0x84
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0x4000b0af00?, {0x5d08388, 0x400836dc20}, 0x4001e7aa08, 0x4001e7a5e8?, 0x4001e7a5e8, 0x40092dc380, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1600 +0x184
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0x4001e7a000, {0x5d08388?, 0x400836dc20}, 0x4001e7a5e8, 0x0?, {0x5d59f80?, 0x400836d800}, 0x4006bfc488?, 0x40043d6c98)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1795 +0x448
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0x4001e7a000, {0x5d082e0, 0x4004cd08c0}, 0x4001e7a5e8, {0x5d59f80, 0x400836d800})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1378 +0x8b8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0x4001e7a000, {0x5d08388?, 0x400836d7a0?}, {{0x5d31520, 0x400a4ead20}, {0x0, 0x0, 0x0}, {0x4b0914c, 0x9f}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:785 +0x1f94
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x5d08388?, 0x400836d7a0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:137 +0x8c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x5d08388?, {0x5d08388?, 0x400836d7a0?}, {0x5d31520?, 0x400a4ead20?}, 0x460d14?, 0x4001caa578?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2683 +0x280
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x4001e7a000, {0x5d08388, 0x400836d7a0}, {{0x5d31520, 0x400a4ead20}, {0x0, 0x0, 0x0}, {0x4b0914c, 0x9f}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:136 +0x36c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x4001caabb8, 0x4001e7a000, 0x4001caa878, 0x4001caa938, 0x4001caa978, 0x400899f958, 0x400899f948, 0x400899f968)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1992 +0x2e0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x4001e7a000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1997 +0xc88
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x4001e7a000, {0x5d08388, 0x400836d380}, 0x400836c930?, 0x10ae420?, 0x4008763000?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1913 +0x18c
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:172 +0x90
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:171 +0x150
```

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1313,6 +1313,15 @@ func (t *logicTest) newCluster(
 		// disable stats collection on system tables in order to have
 		// deterministic tests.
 		stats.AutomaticStatisticsOnSystemTables.Override(context.Background(), &st.SV, false)
+		if t.cfg.UseFakeSpanResolver {
+			// We will need to update the DistSQL span resolver with the fake
+			// resolver, but this can only be done while DistSQL is disabled.
+			// Note that this is needed since the internal queries could use
+			// DistSQL if it's not disabled, and we have to disable it before
+			// the cluster started (so that we don't have any internal queries
+			// using DistSQL concurrently with updating the span resolver).
+			sql.DistSQLClusterExecMode.Override(context.Background(), &st.SV, int64(sessiondatapb.DistSQLOff))
+		}
 		return st
 	}
 	var corpusCollectionCallback func(p scplan.Plan, stageIdx int) error
@@ -1451,8 +1460,14 @@ func (t *logicTest) newCluster(
 
 	t.cluster = serverutils.StartNewTestCluster(t.rootT, cfg.NumNodes, params)
 	if cfg.UseFakeSpanResolver {
-		fakeResolver := physicalplanutils.FakeResolverForTestCluster(t.cluster)
-		t.cluster.Server(t.nodeIdx).SetDistSQLSpanResolver(fakeResolver)
+		// We need to update the DistSQL span resolver with the fake resolver.
+		// Note that DistSQL was disabled in makeClusterSetting above, so we
+		// will reset the setting after updating the span resolver.
+		for nodeIdx := 0; nodeIdx < cfg.NumNodes; nodeIdx++ {
+			fakeResolver := physicalplanutils.FakeResolverForTestCluster(t.cluster)
+			t.cluster.Server(nodeIdx).SetDistSQLSpanResolver(fakeResolver)
+		}
+		serverutils.SetClusterSetting(t.rootT, t.cluster, "sql.defaults.distsql", "auto")
 	}
 
 	connsForClusterSettingChanges := []*gosql.DB{t.cluster.ServerConn(0)}

--- a/pkg/sql/physicalplan/fake_span_resolver.go
+++ b/pkg/sql/physicalplan/fake_span_resolver.go
@@ -167,15 +167,11 @@ func (fit *fakeSpanResolverIterator) Seek(
 	// Build ranges corresponding to the fake splits and assign them random
 	// replicas.
 	fit.ranges = make([]fakeRange, len(splits)-1)
-	// TODO(michae2): Condense this logic when #100051 is fixed.
-	nodes := fit.fsr.nodes
-	n := len(nodes)
 	for i := range fit.ranges {
-		j := fit.rng.Intn(n)
 		fit.ranges[i] = fakeRange{
 			startKey: splits[i],
 			endKey:   splits[i+1],
-			replica:  nodes[j],
+			replica:  fit.fsr.nodes[fit.rng.Intn(len(fit.fsr.nodes))],
 		}
 	}
 

